### PR TITLE
planner/core: refactor `deriveStats` for LogicalPlan

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

For cascades planner, we would keep `LogicalProperty` for each `Group`, and we want to compute it lazily, i.e, after exploration phase, and before implementation phase, we generate the `LogicalProperty`, which includes `StatsInfo`. Current `deriveStats` is a recursive procedure which would pull up stats info of child plan node internally, in cascades planner, we should get child stats from child `Group` instead. To reuse the stats computation logic of each plan node, refactor this part a little.

### What is changed and how it works?

- Add a new interface `DeriveStats` for LogicalPlan to calculate stats info for current plan node according to child stats, and change previous `deriveStats` to `recursiveDeriveStats` which leverages new interface for code re-usage.
- Remove `recursiveDeriveStats` for logical plan nodes, they would reuse `baseLogicalPlan::recursiveDeriveStats` instead.

### Check List 

Tests 

 - Unit test: no code logic change, so existing unit tests are supposed to cover the changes.

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

N/A

Related changes

N/A

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8753)

### CODE REVIEW

**Comment by eurekaka** [comment]
@zz-jason @winoros @lamxTyler PTAL

**Comment by eurekaka** [comment]
@zz-jason @winoros PTAL

File: planner/core/stats.go [hunk 1]
```diff
@@ -25,10 +25,15 @@ func (p *basePhysicalPlan) StatsCount() float64 {
 	return p.stats.RowCount
 }
 
-func (p *LogicalTableDual) deriveStats() (*property.StatsInfo, error) {
+func (p *LogicalTableDual) recursiveDeriveStats() (*property.StatsInfo, error) {
```
Comment (winoros) [inline planner/core/stats.go]: Then it just use `baseLogicalPlan`'s `recursiveDeriveStats` is enough?

File: planner/core/stats.go [hunk 2]
```diff
@@ -40,20 +39,35 @@ func (p *LogicalTableDual) deriveStats() (*property.StatsInfo, error) {
...
 
-func (p *baseLogicalPlan) deriveStats() (*property.StatsInfo, error) {
+func (p *baseLogicalPlan) recursiveDeriveStats() (*property.StatsInfo, error) {
 	if p.stats != nil {
 		return p.stats, nil
 	}
 	if len(p.children) > 1 {
-		panic("LogicalPlans with more than one child should implement their own deriveStats().")
+		err := ErrInternal.GenWithStack("LogicalPlans with more than one child should implement their own recursiveDeriveStats().")
+		return nil, errors.Trace(err)
```
Comment (zz-jason) [inline planner/core/stats.go]: `GenWithStack()` generates an error with the call stack. We can just return that error, no need to call `errors.Trace()`.

File: planner/core/stats.go [hunk 3]
```diff
@@ -40,20 +39,35 @@ func (p *LogicalTableDual) deriveStats() (*property.StatsInfo, error) {
...
 
+// DeriveStats implement LogicalPlan DeriveStats interface.
+func (p *baseLogicalPlan) DeriveStats(childStats []*property.StatsInfo) (*property.StatsInfo, error) {
+	if len(childStats) == 1 {
+		p.stats = childStats[0]
+		return p.stats, nil
+	}
+	if len(childStats) > 1 {
+		err := ErrInternal.GenWithStack("LogicalPlans with more than one child should implement their own DeriveStats().")
+		return nil, errors.Trace(err)
```
Comment (zz-jason) [inline planner/core/stats.go]: ditto

File: planner/core/stats.go [hunk 4]
```diff
@@ -127,30 +139,33 @@ func (ds *DataSource) deriveStats() (*property.StatsInfo, error) {
...
-	}
-	p.stats = childProfile.Scale(selectionFactor)
+// DeriveStats implement LogicalPlan DeriveStats interface.
+func (p *LogicalSelection) DeriveStats(childStats []*property.StatsInfo) (*property.StatsInfo, error) {
+	p.stats = childStats[0].Scale(selectionFactor)
 	return p.stats, nil
 }
 
-func (p *LogicalUnionAll) deriveStats() (*property.StatsInfo, error) {
+func (p *LogicalUnionAll) recursiveDeriveStats() (*property.StatsInfo, error) {
```
Comment (zz-jason) [inline planner/core/stats.go]: Seems this code logic is the same with `baseLogicalPlan`'s? Except that `baseLogicalPlan` only handles one child, while `LogicalUnionAll` handles multi-child.

File: planner/core/stats.go [hunk 5]
```diff
@@ -256,25 +251,32 @@ func (la *LogicalAggregation) deriveStats() (*property.StatsInfo, error) {
...
-// deriveStats prepares property.StatsInfo.
+// recursiveDeriveStats prepares property.StatsInfo.
 // If the type of join is SemiJoin, the selectivity of it will be same as selection's.
 // If the type of join is LeftOuterSemiJoin, it will not add or remove any row. The last column is a boolean value, whose Cardinality should be two.
 // If the type of join is inner/outer join, the output of join(s, t) should be N(s) * N(t) / (V(s.key) * V(t.key)) * Min(s.key, t.key).
 // N(s) stands for the number of rows in relation s. V(s.key) means the Cardinality of join key in s.
 // This is a quite simple strategy: We assume every bucket of relation which will participate join has the same number of rows, and apply cross join for
 // every matched bucket.
-func (p *LogicalJoin) deriveStats() (*property.StatsInfo, error) {
+func (p *LogicalJoin) recursiveDeriveStats() (*property.StatsInfo, error) {
```
Comment (zz-jason) [inline planner/core/stats.go]: ditto

**Comment by zz-jason** [comment]
@eurekaka How about use `baseLogicalPlan.recursiveDeriveStats()` to recursively collect child statistics and call `self.DeriveStats` to complete the statistics derivation for each logical plan node?

**Pull request merged**